kolibri/core/assets/src/api_resource.js

```diff
@@ -0,0 +1,263 @@
+        reject(response);
+      });
+    });
+  }
+  get url() {
+    return this.resource.collectionUrl;
+  }
+
+  /**
+   * Make a model a member of the collection - record in the models Array, and in the mapping
+   * from id to model. Will automatically instantiate Models for data passed in as objects, and
+   * deduplicate within the collection.
+   * @param {(Object|Model|Object[]|Model[])} models - Either an Array or single instance of an
+   * object or Model.
+   */
+  set(models) {
+    let modelsToSet;
+    if (!Array.isArray(models)) {
+      modelsToSet = [models];
+    } else {
+      modelsToSet = models;
+    }
+
+    modelsToSet.forEach((model) => {
+      // Note: this method ensures instantiation deduplication of models within the collection
+      //  and across collections.
+      const setModel = this.resource.addModel(model);
+      if (!this._model_map[model.id]) {
+        this._model_map[model.id] = setModel;
+        this.models.push(setModel);
+      }
+    });
+  }
+}
+
+/** Class representing a single API resource.
+ *  Contains references to all Models that have been fetched from the server.
+ *  Can also be subclassed in order to create custom behaviour for particular API resources.
+ */
+class Resource {
+  /**
+   * Create a resource with a Django REST API name corresponding to the name parameter.
+   * @param {String} name - The Django REST framework name for the API endpoint.
+   * @param {String} idKey - The primary key field for the API endpoint, defaults to 'id'.
+   * @param {Kolibri} kolibri - The current instantiated instance of the core app.
+   */
+  constructor({ name, idKey, kolibri } = { idKey: 'id' }) {
+    this.models = {};
+    this.name = name;
+    this.idKey = idKey;
+    this.kolibri = kolibri;
+  }
+
+  /**
+   * Optionally pass in data and instantiate a collection for saving that data or fetching
+   * data from the resource.
+   * @param {Object[]} data - Data to instantiate the Collection - see Model constructor for
+   * details of data.
+   * @returns {Collection} - Returns an instantiated Collection object.
+   */
+  getCollection(data = []) {
+    return new Collection(data, this);
+  }
+
+  /**
+   * Get a model by id - this will either return a Model that has already been instantiated,
+   * or will instantiate a model
+   * that can later be fetched.
+   * @param {String} id - The primary key of the Model instance.
+   * @returns {Model} - Returns a Model instance.
+   */
+  getModel(id) {
+    let model;
+    if (!this.models[id]) {
+      model = new Model({ [this.idKey]: id }, this);
+    } else {
+      model = this.models[id];
+    }
+    return model;
+  }
+
+  /**
+   * Add a model to the resource for deduplication, dirty checking, and tracking purposes.
+   * @param {Object|Model} model - Either the data for the model to add, or the Model itself.
+   * @returns {Model} - Returns the instantiated Model.
+   */
+  addModel(model) {
+    if (!(model instanceof Model)) {
```

Comment:
I don't think so?

```javascript
  addModelData(data) {
    const model = new Model(model, { resource: this });
    if (!this.models[model.id]) {
      this.models[model.id] = model;
    } else {
      this.models[model.id].set(model.attributes);
    }
    return this.models[model.id];
  }
```

Not a big deal – I just generally like to avoid overloading parameters in JS because the weak typing makes it harder to refactor later.